### PR TITLE
separate thenValue/thenError when rpc fail to handle timeout

### PR DESCRIPTION
**Pull request by critical27**
Separate from #2903, we could recognize time out from exception type:
1. For storage client, don't call `invalidLeader`, it will make further request fail again.
2. For raft client, we should add some back-off time for retry, this will be handled in another PR.

### CODE REVIEW

**Comment by Shylock-Hg** [comment]
How about meta client?

**Comment by critical27** [comment]
> How about meta client?

Meta client already has a retry mechanism, and all rpc in meta should be small enough, there should not be any timeout. I'll add it if necessary. WDYT?

**Comment by Shylock-Hg** [comment]
> > How about meta client?
> 
> Meta client already has a retry mechanism, and all rpc in meta should be small enough, there should not be any timeout. I'll add it if necessary. WDYT?

Got.

File: src/kvstore/raftex/Host.cpp [hunk 1]
```diff
@@ -358,6 +344,43 @@ void Host::appendLogsInternal(folly::EventBase* eb, std::shared_ptr<cpp2::Append
...
+                     }
+                     self->setResponse(r);
+                     self->lastLogIdSent_ = self->logIdToSend_ - 1;
+                   }
+                   // a new raft log or heartbeat will trigger another appendLogs in Host
+                   self->noMoreRequestCV_.notify_all();
+                   return;
+                 })
+      .thenError(folly::tag_t<std::exception>{}, [self = shared_from_this()](std::exception&& ex) {
+        VLOG(2) << self->idStr_ << ex.what();
```
Comment (darionyaphet) [inline src/kvstore/raftex/Host.cpp:373]: The error message should use LOG(ERROR)?

File: src/kvstore/raftex/Host.cpp [hunk 2]
```diff
@@ -358,6 +344,43 @@ void Host::appendLogsInternal(folly::EventBase* eb, std::shared_ptr<cpp2::Append
             return;
           }
         }
+      })
+      .thenError(folly::tag_t<TransportException>{},
+                 [self = shared_from_this(), req](TransportException&& ex) {
+                   VLOG(2) << self->idStr_ << ex.what();
```
Comment (darionyaphet) [inline src/kvstore/raftex/Host.cpp:350]: ditto

**Review by panda-sheep — APPROVED** [review]
Good job！👍

File: src/clients/storage/StorageClientBase-inl.h [hunk 3]
```diff
@@ -138,49 +139,64 @@ folly::SemiFuture<StorageRpcResponse<Response>> StorageClientBase<ClientType>::c
...
-                  auto* leader = code.get_leader();
-                  if (isValidHostPtr(leader)) {
-                    updateLeader(spaceId, code.get_part_id(), *leader);
-                  } else {
-                    invalidLeader(spaceId, code.get_part_id());
-                  }
-                } else if (code.get_code() == nebula::cpp2::ErrorCode::E_PART_NOT_FOUND ||
-                           code.get_code() == nebula::cpp2::ErrorCode::E_SPACE_NOT_FOUND) {
-                  invalidLeader(spaceId, code.get_part_id());
+          .thenValue([this, context, host, spaceId, start](Response&& resp) {
```
Comment (darionyaphet) [inline src/clients/storage/StorageClientBase-inl.h:142]: the difference between `then` and `thenValue` is separate error handling?

Comment (critical27) [reply src/clients/storage/StorageClientBase-inl.h:142]: In general, then = thenTry = thenValue + thenError

**Review by liuyu85cn — APPROVED** [review]
LGTM

**Review by darionyaphet — APPROVED** [review]

**Pull request merged**